[MLv2] Test `metabase.lib.convert` by round-tripping every `mbql-query`

bshepherdson opened the pull request:
This is a great way to make sure the conversion in both directions
is robust; every feature of MBQL is exercised somewhere in the BE tests.

In a few cases the `mbql-query` is called from a context where a
`(is ...)` test won't work - those places can use a `mbql-query-no-test`
instead. There were actually only two of these.

This also includes `metabase.lib.js/legacy-query` for converting from a
CLJS pMBQL query to a JS legacy query; it's the inverse of `lib.js/query`.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29487)
<!-- Reviewable:end -->

bshepherdson committed:
[MLv2] Test `metabase.lib.convert` by round-tripping every `mbql-query`

This is a great way to make sure the conversion in both directions
is robust; every feature of MBQL is exercised somewhere in the BE tests.

In a few cases the `mbql-query` is called from a context where a
`(is ...)` test won't work - those places can use a `mbql-query-no-test`
instead. There were actually only two of these.

bshepherdson committed:
`:aggregation-options` wasn't getting added by `->legacy-MBQL`

Now it works.

bshepherdson committed:
fix some missing cases of `:aggregation-options`

camsaul commented:
*[`resources/sample-database.db.mv.db` line 0 at r2](https://reviewable.io/reviews/metabase/metabase/29487#-NRdA9vfAVHGB_Ehljzd:-NRdA9vfAVHGB_Ehljze:bdok7e7) ([raw file](https://github.com/metabase/metabase/blob/f36bf1e8a88ad48181349529137340dbf7a800ae/resources/sample-database.db.mv.db#L0)):*
I don't think you want to commit changes to this file


<!-- Sent from Reviewable.io -->

camsaul reviewed (approved):
> Reviewed 6 of 8 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29487)* status: all files reviewed, 1 unresolved discussion (waiting on @bshepherdson)



<!-- Sent from Reviewable.io -->

bshepherdson reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29487)* status: all files reviewed, 1 unresolved discussion (waiting on @camsaul)

___
*[`resources/sample-database.db.mv.db` line  at r2](https://reviewable.io/reviews/metabase/metabase/29487#-NRdA9vfAVHGB_Ehljzd:-NRdAcLNAN4CfpGZLAHH:b-igsmkn) ([raw file](https://github.com/metabase/metabase/blob/f36bf1e8a88ad48181349529137340dbf7a800ae/resources/sample-database.db.mv.db#L)):*
<details><summary><i>Previously, camsaul (Cam Saul) wrote…</i></summary><blockquote>

I don't think you want to commit changes to this file
</blockquote></details>

Ah, damn, I missed that that snuck in. I'll fix it after the CI run is done.


<!-- Sent from Reviewable.io -->

bshepherdson committed:
Remove lib.dispatch hack; fix broken Druid test input.

This conversion requires top-level `:aggregation-options`. The schema
suggests these are allowed anywhere, but in practice they're always at
the top level like `:aggregation [[:aggregation-options ...]]` and can't
be nested. (Except in that one Druid test with bad input, now fixed.)

bshepherdson reviewed (commented):
> Dismissed @camsaul from a discussion.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29487)* status: 6 of 11 files reviewed, all discussions resolved (waiting on @camsaul)

___
*[`resources/sample-database.db.mv.db` line  at r2](https://reviewable.io/reviews/metabase/metabase/29487#-NRdA9vfAVHGB_Ehljzd:-NRi9l1jAk9Ce1ey_tyw:b-ybw447) ([raw file](https://github.com/metabase/metabase/blob/f36bf1e8a88ad48181349529137340dbf7a800ae/resources/sample-database.db.mv.db#L)):*
<details><summary><i>Previously, bshepherdson (Braden Shepherdson) wrote…</i></summary><blockquote>

Ah, damn, I missed that that snuck in. I'll fix it after the CI run is done.
</blockquote></details>

Fixed now.


<!-- Sent from Reviewable.io -->

bshepherdson committed:
fix bad merge

camsaul reviewed (changes requested):
This LGTM except for that test suggesting that `:aggregation-options` no longer works inside of something like `:+`, that is technically something that should work and I don't want to break that

> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29487#-:-NRj2vpdC3pJ77KfM_S0:bqtu1cy)* status: all files reviewed, 1 unresolved discussion (waiting on @bshepherdson)

___
*[`modules/drivers/druid/test/metabase/driver/druid/query_processor_test.clj` line 235 at r3](https://reviewable.io/reviews/metabase/metabase/29487#-NRj2VYnCrdQfWqTs6UX:-NRj2VYnCrdQfWqTs6UY:bj4fh7m) ([raw file](https://github.com/metabase/metabase/blob/fc3da056fb633c03cebfdfd217fc5158fdf39011/modules/drivers/druid/test/metabase/driver/druid/query_processor_test.clj#L235)):*
> ```Clojure
>                 :mbql?       true}
>                (query->native
>                 {:aggregation [[:aggregation-options [:+ 1 [:distinct $checkins.venue_name]] {:name "__distinct_0"}]]})))))))
> ```

The old version is theoretically technically legal and should work.


<!-- Sent from Reviewable.io -->

bshepherdson committed:
working with dynamic var

bshepherdson committed:
add defmethods for every expression and aggregation clause

bshepherdson reviewed (commented):
It's working now, with tests to back it up.

> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29487#-:-NRk4lUr6sr0p6QlwB1E:bvvr1p0)* status: 6 of 11 files reviewed, 1 unresolved discussion (waiting on @camsaul)

___
*[`modules/drivers/druid/test/metabase/driver/druid/query_processor_test.clj` line 235 at r3](https://reviewable.io/reviews/metabase/metabase/29487#-NRj2VYnCrdQfWqTs6UX:-NRk4KK787CGlYoX3Dl2:b-krkvka) ([raw file](https://github.com/metabase/metabase/blob/fc3da056fb633c03cebfdfd217fc5158fdf39011/modules/drivers/druid/test/metabase/driver/druid/query_processor_test.clj#L235)):*
<details><summary><i>Previously, camsaul (Cam Saul) wrote…</i></summary><blockquote>

The old version is theoretically technically legal and should work.
</blockquote></details>

I've reverted the test and changed how `->legacy-MBQL` handles it.


<!-- Sent from Reviewable.io -->

camsaul reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29487#-:-NRnX8Pz5Lynhm7L_pSk:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @bshepherdson)



<!-- Sent from Reviewable.io -->

bshepherdson committed:
normalize before round trip to fix any ancient MBQL oddities

bshepherdson committed:
Default handling for maps in `->legacy-MBQL`
